### PR TITLE
Offline support - v1

### PR DESCRIPTION
**Pull request by vishalnarkhede**
### Storage options

This sdk has added support to use the chat in offline mode.
We support two options for storage:

1. **AsyncStorage - https://github.com/react-native-community/async-storage**

   #### Install

   Please follow instructions mentioned here - https://github.com/react-native-community/async-storage#install

   #### Usage

   ```
   import {
     ...
     LocalStorage,
   } from 'stream-chat-react-native';

   import AsyncStorage from '@react-native-community/async-storage';

   ....
   storage = new LocalStorage(chatClient, AsyncStorage, 'async-storage');

   ...

     render() {
       return (
         <SafeAreaView>
           <Chat client={chatClient} storage={storage}>
             ...
           </Chat>
         </SafeAreaView>
       );
   ```

   Following methods are exposed for our Async Storage interface:

   - `deleteAll()` Deletes all the stored data related to stream chat.
   - `getAllKeys()` Returns async storage keys for stream chat.

   e.g.,

   ```
   // To clear the local database related to stream chat
   storage = new LocalStorage(chatClient, AsyncStorage, 'async-storage');
   storage.db.deleteAll();
   ```

2. **RealmJS - https://github.com/realm/realm-js**

   #### Install

   Please follow the instructions given here - https://realm.io/docs/javascript/latest/#installation

   #### Usage

   ```
   import {
     ...
     LocalStorage
   } from 'stream-chat-react-native';

   const Realm = require('realm');
   ....
   storage = new LocalStorage(chatClient, Realm, 'realm');
   ...

     render() {
       return (
         <SafeAreaView>
           <Chat client={chatClient} storage={storage}>
             ...
           </Chat>
         </SafeAreaView>
       );
   ```

   Following methods are exposed for our Async Storage interface:

   - `deleteAll()` Deletes all the stored data related to stream chat.

   #### Limitations

    - RealmJS doesn't work with [Hermes engine](https://hermesengine.dev). For more details, take a look at this issue - https://github.com/realm/realm-js/issues/2455
    - Expo does not support Realm. - https://expo.canny.io/feature-requests/p/support-for-realm

### Images

Images are not cached/stored by default. Although it is totally supported.
We support `ImageComponent` prop for `ChannelList` and `Channel` component, which will be
used to display images in places components like avatar and attachments.

You can use [`CachedImage`](https://www.npmjs.com/package/@stream-io/react-native-cached-image) for this purpose.
Its a package, which is forked from https://github.com/fungilation/react-native-cached-image.
It handles storing images on device and using local images whenever required.

Ofcourse, you don't need to use it, you can also write your own component for this purpose.

#### Installing

```
yarn add @stream-io/react-native-cached-image rn-fetch-blob

// For RN > 0.60
cd ios && pod install && cd ..
// For RN < 0.60
react-native link rn-fetch-blob
```

#### Usage

```
import {
  CachedImage,
  ImageCacheProvider
} from 'react-native-cached-image';

...

<ChannelList
  filters={filters}
  sort={sort}
  options={options}
  ImageComponent={CachedImage}
/>
```

### Logging

To log the errors, you can set the logger function on storage class:

```
storage.setLogger((storageType, msg, extraData) => {
  reactotron.log(storageType, msg, extraData);
})
```

**TODO List**

**Done**

✅ Sync channels in local storage when queryChannels is called
✅ When ChannelList component is loaded, first query the channels from local storage and display it immediately before waiting for results from remote server
✅ When new message is received/deleted/updated, sync it in local storage
✅.When new reaction is received/deleted/updated, sync it in local storage
✅ When new member is received/deleted/updated, sync it in local storage
✅ When read state of channel is updated, sync it with local storage
✅ Tested and iOS and android both (for RN 0.60). Works fine :)
✅ Revisit the resync case
✅ Add a message to local storage when message is sent in offline mode
✅ Publish npm package for `react-native-cached-images` from fork, for our use
✅ Handle failures

**Pending**

- Test on RN < 0.60

Related request: https://github.com/GetStream/stream-chat-react-native/issues/51

### CODE REVIEW

**Comment by aniciom** [comment]
I think its worth to mention this issue https://github.com/realm/realm-js/issues/2455 for RealmJS for those whose are using Hermes Engine, otherwise the app would end up crashing.

Thanks

File: package.json [hunk 1]
```diff
@@ -69,6 +69,7 @@
     "react-art": "^16.8.6",
     "react-native-actionsheet": "^2.4.2",
     "react-native-image-zoom-viewer": "^2.2.25",
+    "react-native-cached-image": "link:../react-native-cached-image",
```
Comment (vishalnarkhede) [inline package.json]: Need to publish forked package of this repo. Original repo is not React-native 0.60 compatible

File: expo-package/src/index.js [hunk 2]
```diff
@@ -3,9 +3,28 @@
  * And they have been removed from Expo package. But these packages don't work with Expo 32 sdk (because of linking issue).
```
Comment (vishalnarkhede) [inline expo-package/src/index.js]: Ignore this for now!!

File: native-package/src/index.js [hunk 3]
```diff
@@ -80,6 +83,7 @@ registerNativeHandlers({
       };
     }
   },
+  storage: new RealmStorage(),
```
Comment (vishalnarkhede) [inline native-package/src/index.js]: Not relevant anymore!!

File: src/storage/realmStorage/index.js [hunk 4]
```diff
@@ -0,0 +1,291 @@
...
+   * Store channels in database
+   * @param {*} query
+   * @param {*} channels
+   * @param {*} resync
+   */
+  async storeChannels(query, channels, resync) {
+    const realm = await this.getRealm();
+
+    realm.write(() => {
+      if (resync) {
```
Comment (vishalnarkhede) [inline src/storage/realmStorage/index.js]: This needs to go away!!

**Comment by vishalnarkhede** [comment]
Moved to https://github.com/GetStream/stream-chat-react-native/pull/306

**Pull request closed without merging**